### ISSUE TITLE
Explain how hiding will actually work

### ISSUE THREAD

**Issue by FrameMuse** [domenic/proposal-function-prototype-tostring-censorship#52]
Maybe I didn't notice it in the `README.md`, but I don't really understand how this will help to conceal sensitive code if you either get the packages via `npm i`, which retrieves [npmjs.com](https://www.npmjs.com/) or you use some kind of CDN services directly on a webpage, which also retrieves a remote service.

**How is hiding possible while already having all the code in hand?**

**Comment by ljharb** [comment]
It’s not about concealing code from humans, it’s about concealing it from JavaScript.

**Comment by FrameMuse** [comment]
Isn't this part about concealing a code from humans?

> The "sensitive" directive is intended to be expanded in the future as new security-impacting information leakages are discovered or added to the language.

Also this one

> They (callers) can also extract secret values from the source text to compromise an application's attempts at encapsulation.

_From `README.md`_

How does hiding an implementation will help to resolve this problem if a source code is always open?

My concern is that by removing one convenient way of retrieving certain function will not solve the problem because people, who actually wants to do so, will anyway be able to access functions by creating their own tools or changing browser behavior to reveal hidden code after all.

I understand having this proposal (which can be revealed by using a revealing method - #42) to help implementers to outline that this code can be changed in future or shouldn't be relied on. But I don't see this proposal as sensitive information concealment or a security measure.

**Comment by ljharb** [comment]
Because the attack vector isn't from humans reading the code, it's from code attacking or exfiltrating. "callers" means "javascript functions", not "humans reading things".

**Comment by FrameMuse** [comment]
> Because the attack vector isn't from humans reading the code, it's from code attacking or exfiltrating. "callers" means "javascript functions", not "humans reading things".

Yeah, make sense, though I'm talking more about sources, theoretically you can write a function that will extract source code without using `.toString` because for web pages the only way to load resources - using `script` tag, no one will restrict that.

As for server-side I'm not sure if the proposal is useful. I can't imagine how something can access a server and inject something in code, usually there is no active console for that and if something have access to a server, it's already late.

From the side of libraries, usually they are open or scoped to an organization you work on, so if something's wrong, you know who you should blame. I doubt open libraries would sacrifice their reputation to inject something in your code.

From the side of viruses, they may attack your browsers, but I doubt they would target specific websites, usually they just stealing your account creds and that's enough. But again, if we have close one way of attacking (that eventually will harm debugging), we come to the beginning of this comment.

I've read something about probable optimization because of hiding a _function string representation_. If it would be so, we should think about always hiding _function string representation_ and expose it only for debug purposes (i.e. errors and logging).

---

Maybe I got your point wrong, please unfold more what you mean by saying "attack" and where it may come from.

**Comment by ljharb** [comment]
Servers run untrusted third party code all the time, as do browsers. Open source libraries have many documented incidents of bad actors hijacking a package and sneaking in malicious code.

Either way, the readme explains the primary motivation - anything that's observable about your API risks becoming part of it, including the string representation of your functions. They can't just be hidden by default, but this proposal provides a way for a function author to opt in to doing it.

**Comment by c-ridgway** [comment]
I dislike this feature for a reason stated above. 

Many frameworks will just blindly implement this, even browsers - we realy on these exploits to make things happen. If you're running code on your own machine, it's up to them to make sure system security is in place. I hope this feature doesn't make it.

**Comment by ljharb** [comment]
Relying on exploits to make things happen is exceedingly unwise. If you need a feature, convince the framework or library to add it - or else, you simply don't get to have that feature.

**Comment by FrameMuse** [comment]
@ljharb I agree relying on exploits is not a good idea at least because they may be closed. Though the problem that @c-ridgway states is that one JS code can rely on internal behavior of another JS code - not from `toString`, but because a **Person** (behind the first one) has peeked a source code of the 2nd one.

1. As for **Security** - this will never help, and as we already discussed, it's not made to cover any security concerns.
2. As for **Hiding** for Humans - as @ljharb mentioned many times, it's not the purpose and the `README.md` also states that "Devtools won't be affected".
3. As for **Hiding** code from JS itself - I already mentioned that it simply won't work for browsers, as the only method to deliver a script to a page - via `script` element (correct me if I don't know smth) and any script can request every `script` element on the page, which raises concern how the code is actually concealed, is it going to conceal fetched scripts with "sensitive" word too? As for non-browser environment, it's often that script can access files directly, which again defeats the purpose of source concealment (from JS itself) - the library author somehow must know that it will be used in an environment without any potential source code exposure. But if a developer really cares about code concealment I don't think they will rely on "sensitive" tag only. If one needs to make sure that script they use is not exposing underlying sources, there are already ways to do this, e.g. by wrapping a function into another function and minifying, which works great.

In short: the goal of the concealment is unclear. There is no impactful use-cases stated in the `README.md`, except releasing memory of held source strings, but it's excluded from this proposal.

There is only one example stated: "Error Stack Trace" where `at bound (domain.js:420:14)` gets removed, which is completely insufficient as an example - who would exchange debug ability for some uncertain "implementation hiding". That is said, it's absolutely unclear HOW the source code is supposed to be hidden if there are other ways to peek it **in JS itself**. In many cases, the source code is minified/uglified, which makes those references like `asdasd.js:420:14` useless as they are not applicable. If the source code is already known for JS, a script can be written that will do the same job as `.toString` because the source is very likely still accessible in the environment, even if it's hidden from the language itself.

Usually even if some code relies on exploits, it's not because some JS can see internal implementation of an external function, it's because a developer had intent to do so - if it was there beforehand, it will stay even if this proposal is shipped. The developer anyway would continue doing as it was, but now with a different approach. So what it addresses then?

It's stated that it's for developers, because they may want to hide their sources - but no one yet answers "why they may want that". If it's not protecting a human from **relying** on it nor human from **reading** it, then **why** we need this kind of hiding at all? It would harm the ability to debug much more than any other advantage that it might provide.
